### DISCUSSION TITLE
Shell integration prevents some apps from launching with ghostty

### DISCUSSION THREAD

**Discussion by andrejdaskalov** [ghostty-org/ghostty#2957] [Feature Requests, Ideas] [answered]
When launching an app through a .desktop file, it may fail due to ghostty adding --posix flag for bash shell integration.
![image](https://github.com/user-attachments/assets/71441a6d-222a-443a-a86e-845041a75190)

In this case, the app being launched is neovim through the app menu, while the default terminal emulator is set to Ghostty. 
This appears to be where the issue is: (src/termio/shell_integration.zig)
```zig
   // Start accumulating arguments with the executable and `--posix` mode flag.
    if (iter.next()) |exe| {
        try args.append(exe);
    } else return null;
    try args.append("--posix");
```
Perhaps this needs to be rethought?

EDIT: I have bash shell integration set in the config.

**Comment by jparise** [comment]
It seems like we shouldn't be using the shell integration path at all if we're spawning a non-shell executable. Or am I misunderstanding?

**Reply by jparise** [reply]
Oh, scratch that. I had the wrong idea about how Ghostty was being launched.

I think it's more a matter of the current code adding `--posix` before we recognize the shell's `-c` option:

https://github.com/ghostty-org/ghostty/blob/0d0d06e50a1a5c5074fa17f0f7c7852c7db5d6b0/src/termio/shell_integration.zig#L191-L194

I can take a look at this soon.

**Reply by andrejdaskalov** [reply]
I could escalate this into an issue then?

**Reply by jparise** [reply]
Actually, I think I was right the first time. Do you have `shell-integration = bash` set in your configuration?

I think the problem is that "forced" shell integration (via that explicit configuration option) skips all of the normal detection mechanisms and we end up assuming that `command = /usr/bin/nvim` (in this case) _is_ a e.g. `bash` executable which can receive our shell integration arguments + environment.

**Reply by andrejdaskalov** [reply]
Yes I do have bash shell integration set. Probably should've had that in the post.

**Reply by jparise** [reply]
I'm going to start by taking a look at getting `shell-integration = detect` (the default) working for bash again. (I disabled it in #1823 to work-around old bash versions; in particular, the one shipped with macOS.)

If I can't get that to work, I think we might need to differentiate a config-set `command` from a command-line-provided (`-e`) command. I think we should only consider the former (`command`) when applying/forcing shell integration.

**Reply by mitchellh** [reply]
> I'm going to start by taking a look at getting shell-integration = detect (the default) working for bash again. (I disabled it in https://github.com/ghostty-org/ghostty/pull/1823 to work-around old bash versions; in particular, the one shipped with macOS.)

I think this would be ideal. I think there's a couple things we can do here.

(1) If `bash` is detected, we can run `bash --version` and look for 4+ to auto-inject. This is very slow but doing it only for bash users and then adding some fast paths (below) might make it work.

(2) For macOS, if its `/bin/bash`, we can assume its Bash v3 because the system directories (i.e. `/bin`) are not writeable without disabling SIP which is extremely rare. For the rare macOS user that disables SIP _and uses default bash_ (???), we can just document they should explicitly opt in.

(3) Maybe instead of (1), we can detect the bash version in the shell integration itself? That might be faster than sub processing `bash --version`, something like this (from ChatGPT):

```bash
#!/bin/bash

# Check the Bash version
if [[ "${BASH_VERSINFO[0]}" -lt 4 ]]; then
    echo "This script requires Bash version 4 or higher."
    exit 1
else
    echo "You are using Bash version $BASH_VERSION."
fi
```

**Reply by jparise** [reply]
> (3) Maybe instead of (1), we can detect the bash version in the shell integration itself? That might be faster than sub processing `bash --version`, something like this (from ChatGPT):

I think the problem we hit previously was that older versions of bash weren't loading our script at all via `--posix` mode, so I wasn't able to use this (much simpler) in-script approach.

If I'm not able to solve that, I like your other suggestions.

**Reply by jparise** [reply]
It took some spelunking, but it appears that Apple's "fork" of bash disables the `--posix` flag:

https://github.com/apple-oss-distributions/bash/blob/e5397a7e74633a4e84194a6c6b609e04077da6f8/bash-3.2/shell.c#L1112-L1114

This appears to have come in as part of an upstream merge:

https://github.com/apple-oss-distributions/bash/commit/3684f6230c86fa451523da13e3ddf2349a2030bc

... but I don't know what the ["bash-76.1"](https://github.com/apple-oss-distributions/bash/releases/tag/bash-76.1) tag represents, and I don't see this code in `bash-3.2` (or `bash-4.0`) itself:

- https://github.com/bminor/bash/blob/bash-3.2-beta/shell.c
- https://github.com/bminor/bash/blob/bash-4.0/shell.c

So, that means that we can't use our POSIX `ENV`-based to shell integration injection with Apple's `/bin/bash` shipped with macOS. Non-Apple Bash 3.2 works fine, so we need the "macOS /bin/bash" check (2) rather than a version-based check. I'll take that approach.

**Reply by jparise** [reply, marked as answer]
#2966 should address the bash shell detection issue, making `shell-integration = bash` unnecessary for most users.

I think there's still further work to be done here, though, because there are valid reasons why someone might force bash using `shell-integration = bash`, and that appears to interfere with the `.desktop` use case described here.

**Reply by jparise** [reply]
And #2971 should address the other part of this problem, where we'd (unexpectedly) attempt to inject shell integration into the `initial-command` (`-e`) execution environment.

**Reply by andrejdaskalov** [reply]
I can confirm after pulling the commits after they were merged, and removing the explicit config option for bash, that everything seems to work as expected now! Shell integration seems to be working fine, and I can launch neovim from launcher just fine. 
I also find the fact that bash users will have a much better zero configuration experience as a great bonus.

**Reply by jparise** [reply]
Great! The things I learned above about Apple's `bash` were the key to unlocking this.

And thanks for starting this discussion!

**Reply by andrejdaskalov** [reply]
No problem! Likewise, I learned a lot from this too.
